README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="./man/figures/CellSurvAssay.png" width="200%" height="200%" />

# CellSurvAssay

<!-- badges: start -->
<!-- badges: end -->

**CellSurvAssay** consists of a couple of tools that can be used to
perform Clonogenic Survival Analysis in R very easily and efficiently.
These two tools are:

-   **CellSurvAssay R package**: This helps even beginner R users to
    perform the analysis in R, while maintaining the flexibility of a
    package.

-   **CellSurvAssay Shiny app**: This is a web application that helps
    users with no experience in R to perform the analysis, in R. The app
    is based on the CellSurvAssay R package and can be accessed
    [here](https://pickeringlab.shinyapps.io/CellSurvAssay-App).

# Purpose of the CellSurvAssay R package

-   This R package has been written around
    [CFAssay](https://bioconductor.org/packages/release/bioc/html/CFAssay.html),
    another R package that can be used to perform Cell Survival Assay
    analysis in R. However, **CellSurvAssay** has it’s own purposes and
    advantages:
    -   it makes performing Clonogenic Survival Analysis in R incredibly
        user-friendly and efficient, even for beginner R users who don’t
        have the luxury of time to dig deeper into R,
    -   it arranges all the commonly used steps of clonogenic assay
        analysis in one location and automates the data wrangling steps
        to the extent that only single lines of code suffice for each
        step of the analysis,
    -   it utilizes `ggplot()` to plot the cell survival curves, and
        builds better quality figures than other available R packages,
    -   it is less time consuming and more convenient for the user, as
        it accepts the raw data for the analysis and calculates the
        plating efficiencies by itself, unlike many automated software
        commonly used,
    -   it offers various method options for parameter estimation and
        calculating plating efficiencies, unlike most other available
        software tools, and
    -   as R is being utilized, the methodology stays open and the
        results reproducible.

------------------------------------------------------------------------

*Below is just a quick workflow that can be used to perform Clonogenic
Survival Analysis using this package. For more details on the functions,
different method options for parameter estimation and calculation of
plating efficiency, and customization of the figures, please refer to
the package vignette.*

------------------------------------------------------------------------

# Installing the package

-   The package is being shared through [GitHub](https://github.com/).
    Running the following set of codes installs and loads the package in
    R memory:

``` r
# if installing from Bioconductor
# install BiocManager, if required
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# install CellSurvAssay
BiocManager::install("CellSurvAssay")
# load CellSurvAssay in R
library(CellSurvAssay) 



# if installing from GitHub
# install devtools, if required
if(!require(devtools)) {
    install.packages("devtools")
    library(devtools)
}
# install CellSurvAssay
install_github("arunangshu-github/CellSurvAssay", 
               build_vignettes = TRUE,
               dependencies = TRUE)
# load CellSurvAssay in R memory
library(CellSurvAssay)  
```

-   This vignette can be accessed from R using the following command:

``` r
browseVignettes("CellSurvAssay")
```

# Importing the data set

-   The function `importData()` helps import the data set.

``` r
datatab <- importData("path/to/file", "type of file")
```

-   The package contains a data set that can be used to get familiar
    with the package. It can be imported in the R environment as below:

``` r
datatab <- CASP8_data
```

# Fitting the Linear Quadratic Model

-   The `lqmodelFit()` function helps fit the linear quadratic model for
    any cell type present in the imported data.

``` r
lqmodelFit(datatab, "shCASP8-N")
#> ****** Cell type: shCASP8-N ******
#> 
#> *** Coefficients of LQ-model for cell survival *** 
#> method = ml 
#> PEmethod = fit 
#> 
#> Logarithmic plating efficiencies PE fitted as intercepts 
#> see remark in the manual, 1.2 
#>      Estimate Std. Error   t value     Pr(>|t|)
#> PE1 -1.238340 0.05429647 -22.80702 7.892816e-08
#> PE2 -1.205679 0.05371585 -22.44550 8.815030e-08
#> PE3 -1.297141 0.05537422 -23.42499 6.559791e-08
#> 
#> Shape parameters alpha and beta 
#>          Estimate  Std. Error    t value    Pr(>|t|)
#> alpha -0.01613085 0.038556758 -0.4183664 0.688215852
#> beta  -0.03678049 0.007020955 -5.2386733 0.001200996
#> 
#> Observed and fitted plating efficiencies (%): 
#>     Experiment   PE PEfitted
#> PE1          1 29.6     29.0
#> PE2          2 31.0     29.9
#> PE3          3 26.7     27.3
#> 
#> Residual Deviance: 9.996061 
#> Total residual sum of weighted squares rsswTot: 10.28712 
#> Residual Degrees of Freedom: 7 
#> Dispersion parameter: 1.469589 
#> 
#> Fraction rssw of rsswTot per Experiment 
#>   Experiment rssw perCent
#> 1          1 1.47    14.3
#> 2          2 1.08    10.5
#> 3          3 7.74    75.3
#> 
#> *** Analysis by CellSurvAssay v0.99.0 ***
```

# Plotting Cell Survival curves

-   The cell survival curves can be plotted using two different
    functions: `plotCSCurve()` and `ggplotCSCurve()`.
-   While the former gives the standard curves provided by the CFAssay
    package, the latter is preferred as it uses the widely popular
    `ggplot()` function of R to plot the curves, allowing all the
    customizations and better graphics. It also allows to easily
    download the figures plotted in the users’ own specifications.

### Individual curves

``` r
ggplotCSCurve(datatab, "shCASP8-NT")
```

![](man/figures/README-unnamed-chunk-8-1.png)<!-- -->

### Multiple curves

``` r
ggplotCSCurve(datatab, "shCASP8-NT", "shCASP8-B", "shCASP8-B+Z", "shCASP8-B+Z+N")
```

![](man/figures/README-unnamed-chunk-9-1.png)<!-- -->

# Comparing two curves

-   The `compareCurves()` function helps us statistically compare two
    curves. It prints the ANOVA results.
-   The null hypothesis is that the parameters
    ![\\alpha](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Calpha "\alpha")
    and
    ![\\beta](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cbeta "\beta")
    of both the models are independent of the two curves, while the
    alternate hypothesis is that the parameters are different.

``` r
compareCurves(datatab, "shCASP8-N", "shCASP8-B+Z+N")
#> ****** Cell type 1: shCASP8-N   ||   Cell type 2: shCASP8-B+Z+N ******
#> 
#> Overall comparison test for coefficients alpha and beta of LQ-models 
#> ==================================================================== 
#> method = ml 
#> PEmethod = fit 
#> 
#> 6 PEs fitted as intercepts. To look at, use simple R print function. 
#> Null hypothesis (Model 1): one set of shape parameters alpha and beta for all data 
#> ---------------------------------------------------------------------------------- 
#>          Estimate  Std. Error    t value     Pr(>|t|)
#> alpha -0.01910478 0.020810067 -0.9180547 3.722228e-01
#> beta  -0.03692732 0.003798381 -9.7218568 4.061270e-08
#> 
#> Goodness-of-fit values 
#>  Residual Deviance: 12.85619 
#>  Total sum of squared weighted residuals rsswTot: 13.19573 
#>  Residual Degrees of Freedom: 16 
#>  Dispersion parameter: 0.8247333 
#> 
#> Alternative hypothesis (Model 2): two sets of shape parameters alpha and beta 
#> ----------------------------------------------------------------------------- 
#>                              Estimate  Std. Error    t value     Pr(>|t|)
#> alpha:curvesshCASP8-B+Z+N -0.02224331 0.031717703 -0.7012901 4.946148e-01
#> alpha:curvesshCASP8-N     -0.01613085 0.030528750 -0.5283823 6.055094e-01
#> beta:curvesshCASP8-B+Z+N  -0.03710309 0.005804313 -6.3923312 1.675728e-05
#> beta:curvesshCASP8-N      -0.03678049 0.005559103 -6.6162636 1.158171e-05
#> 
#> Goodness-of-fit values 
#>  Residual Deviance: 12.60644 
#>  Total sum of squared weighted residuals rsswTot: 12.89855 
#>  Residual Degrees of Freedom: 14 
#>  Dispersion parameter: 0.9213253 
#> 
#> Analysis of Variance Table and F-test
#> Model 2 versus Model 1
#>   Resid. Df Resid. Dev Df Deviance      F Pr(>F)
#> 1        16     12.856                          
#> 2        14     12.606  2  0.24975 0.1355 0.8744
#> 
#> *** Analysis by CellSurvAssay v0.99.0 ***
```

# Calculating Dose Enhancement Ratio

-   The function `calculateDER()` calculates the Dose Enhancement Ratio.

``` r
calculateDER(datatab, "shCASP8-NT", "shCASP8-N", 0.25)
#> *** Dose Enhancement Ratio ***
#> 
#> control = shCASP8-NT
#> treatment = shCASP8-N
#> survival fraction = 0.25
#> method = ml
#> PEmethod = fit
#> DER = 0.945622065301553
#> 
#> *** Analysis by CellSurvAssay v0.99.0 ***
```

-   Here, the DER of shCASP8-NT : shCASP8-N is 0.9456221.


inst/NEWS.md
# CellSurvAssay 0.99.0

## Submission to Bioconductor

* The package is submitted to Bioconductor for review.


R/ggplotCSCurve.R
#' Plots Cell Survival Curves using \code{ggplot()}
#'
#' This function helps to plot the cell survival curves using \code{ggplot()},
#' and is the recommended way to plot this curves over
#' \code{\link{plotCSCurve}}. A single function plots the curves, both for a
#' single or multiple cell type(s), and allows most of the options necessary to
#' customize the figures for publication.
#'
#' @param data A data frame containing at least the following five columns:
#'   \code{"cline"}, \code{"Exp"}, \code{"dose"}, \code{"ncells"}, and
#'   \code{"ncolonies"}.
#' @param ... The name of the cell type(s)/group(s). If entering multiple cell
#'   types, separate each by a comma.
#' @param method Method used for the fit. It's \code{"ml"} (maximum likelihood)
#'   by default. Can be \code{"ls"} (least squares) or \code{"franken"}
#'   (weighted least squares as described by Franken eta al.(2006)).
#' @param PEmethod Controls the value of the plating efficiencies. The default
#'   method, \code{"fit"}, calculates fitted plating efficiencies as model
#'   parameters, while \code{"fix"} uses fixed ones calculated from the observed
#'   zero dose data.
#' @param colors A vector of strings denoting the colors of the curves. Size of
#'   the vector should be equal to or more than the number of cell types
#'   entered. If no colors are chosen by the user, the function chooses them by
#'   random and they might change with each plotting.
#' @param title A string denoting the title of the plot. Default is \code{""}.
#' @param title_size An integer denoting the font size of the title. Default is
#'   \code{20}.
#' @param title_color A string denoting the color of the font of the title.
#'   Default is \code{"black"}.
#' @param title_face A string denoting the type of font face. Accepts
#'   \code{"bold"}, \code{"italic"}, or \code{"bold.italic"}. Default is
#'   \code{"bold"}.
#' @param title_align A string denoting the justification (alignment) of the
#'   plot title. Accepts \code{"center"}, \code{"left"}, or \code{"right"}.
#'   Default is \code{"center"}.
#' @param subtitle A string denoting the subtitle of the plot; default is
#'   \code{""}.
#' @param sub_size,sub_color,sub_face,sub_align Used to customize the font size,
#'   color, face, and alignment of the subtitle respectively, like the plot
#'   title. Default values of these are \code{15}, \code{"black"},
#'   \code{"italic"}, and \code{"center"} respectively.
#' @param xlab A string denoting the label of X-axis; default is \code{"Dose
#'   (Gy)"}.
#' @param xlab_size,xlab_color,xlab_face Used to customize the font size, color,
#'   face, and alignment of the label of the X-axis respectively, like the plot
#'   title. Default values of these are \code{16}, \code{"black"}, and
#'   \code{"bold"} respectively.
#' @param ylab A string denoting the label of y-axis; default is
#'   \code{"Surviving Fraction"}.
#' @param ylab_size,ylab_color,ylab_face Used to customize the font size, color,
#'   face, and alignment of the label of the Y-axis respectively, like the plot
#'   title. Default values of these are \code{16}, \code{"black"}, and
#'   \code{"bold"} respectively.
#' @param legend_title A string denoting the title of the legend; default is
#'   \code{""}.
#' @param ltitle_size,ltitle_color,ltitle_face,ltitle_align Used to customize
#'   the font size, color, face, and alignment of the legend title respectively,
#'   like the plot title. Default values of these are \code{20}, \code{"black"},
#'   \code{"bold"}, and \code{"left"} respectively.
#' @param ltext_size,ltext_color,ltext_face To change the font size, color, and
#'   face of the legend texts. Default values of these are \code{18},
#'   \code{"black"}, and \code{"bold"} respectively.
#' @param xtext_size,xtext_color,xtext_face To change the font size, color, and
#'   face of x axis ticks.Default values of these are \code{14}, \code{"black"},
#'   and \code{"bold"} respectively.
#' @param ytext_size,ytext_color,ytext_face To change the font size, color, and
#'   face of y axis ticks. Default values of these are \code{14},
#'   \code{"black"}, and \code{"bold"} respectively.
#' @param legend_pos A string denoting the position of the legend.
#'   \code{"inside"} is default, and places the legend at the bottom left of the
#'   figure. Also accepts \code{"outside"} (places the legend outside the figure
#'   on the right) and \code{"none"} (removes the legend).
#' @param legend_back A string denoting the color of the legend background.
#'   Default is \code{""}.
#' @param legend_border A string denoting the color of the legnd border. Default
#'   is \code{""}.
#' @param legend_border_width An integer denoting the width of the legend
#'   border. Default is \code{0}.
#' @param rem_minor_x If \code{"no"}, keeps the minor grids in the X-axis.
#'   Default is \code{"yes"}.
#' @param rem_minor_y If \code{"no"}, keeps the minor grids in the Y-axis.
#'   Default is \code{"yes"}.
#' @param emph_major_x If \code{"yes"}, allows to emphasize and customize the
#'   major grids in the X-axis. Default is \code{"no"}.
#' @param emph_major_y If \code{"yes"}, allows to emphasize and customize the
#'   major grids in the Y-axis. Default is \code{"no"}.
#' @param major_x_col,major_x_width Allows the change of the color and width of
#'   the major grids in the X-axis, when \code{emph_major_x = "yes"}. Default
#'   values of these are \code{"gray"} and \code{1} respectively.
#' @param major_y_col,major_y_width Allows the change of the color and width of
#'   the major grids in the Y-axis, when \code{emph_major_y = "yes"}. Default
#'   values of these are \code{"gray"} and \code{1} respectively.
#' @param point_shape An integer denoting the shape of the points in the plot.
#'   Default is \code{16}.
#' @param point_size An integer denoting the size of the points in the plot.
#'   Default is \code{3.5}.
#' @param segment_type An integer denoting the type of the straight line
#'   segments in the plot. Default is \code{1}.
#' @param segment_width An integer denoting the width of straight line segments
#'   in the plot. Default is \code{1}.
#' @param curve_width An integer denoting the width of curved line segments in
#'   the plot. Default is \code{1.5}.
#' @param curve_type An integer denoting the type of curved line segments in the
#'   plot. Default is \code{1}.
#' @param theme The name of the theme to be used in the plot. It should be
#'   preceded by the name of the package in the form: \code{package::theme()}.
#'   Default is \code{ggplot2::theme_bw()}.
#' @param ylim A vector denoting the limits of the y-axis. Defaults are chosen
#'   automatically by \code{ggplot()}.
#' @param ybreaks A vector denoting the y-axis breaks/ticks. Defaults are chosen
#'   automatically by \code{ggplot()}.
#' @param save To save the plot. \code{save = "yes"} saves the plot in the
#'   specified path and format.
#' @param save_path Path where the plot is to be saved. Default is the current
#'   working directory obtained by \code{getwd()}.
#' @param save_filename File name with extension for the plot. Accepts most of
#'   the common extensions, including \code{.pdf}, \code{.jpeg}, and
#'   \code{.png}.
#' @param plot_height A number denoting the height of the plot. Default is
#'   \code{4}.
#' @param plot_width A number denoting the width of the plot. Default is
#'   \code{5}.
#' @param units A string denoting the units of the height and width of the plot
#'   mentioned. Accepts \code{"in"}, \code{"cm"}, \code{"mm"}, or \code{"px"}.
#'   Default is \code{"in"}.
#' @importFrom magrittr %>%
#' @return A \code{ggplot} object.
#' @examples
#' datatab <- CASP8_data
#'
#' # Single curve
#'
#' ggplotCSCurve(datatab, "shCASP8-NT")
#'
#' ggplotCSCurve(datatab, "control-B", colors = "black", xlab = "X-axis", ylab = "Y-axis",
#' title = "Single Plot", point_shape = 1, point_size = 2, segment_width = 1, segment_type = 1,
#' curve_width = 1, curve_type = 1, legend_title = "Cell types", ylim = c(0.008, 1.05),
#' ybreaks = c(0.01, 0.05, 0.20, 0.50, 1), theme = ggplot2::theme_classic())
#'
#' # Multiple Curves
#'
#' ggplotCSCurve(datatab, "shCASP8-NT", "shCASP8-B", "shCASP8-B+Z", "shCASP8-B+Z+N")
#'
#' ggplotCSCurve(datatab, "shCASP8-NT", "shCASP8-B", "shCASP8-B+Z", "shCASP8-B+Z+N",
#' colors = c("darkgreen", "green", "orange", "magenta"), xlab = "X-axis", ylab = "Y-axis",
#' title = "Multiple Curves", point_shape = 15, point_size = 1, segment_width = 1, segment_type = 1,
#' curve_width = 1, curve_type = 1, legend_title = "Cell types", theme = ggplot2::theme_dark(),
#' legend_back = "white", legend_border = "white", legend_border_width = 0.5)
#'
#' \dontrun{
#' # Save a plot
#' ggplotCSCurve(datatab, "shCASP8-NT", "shCASP8-B", "shCASP8-B+Z", "shCASP8-B+Z+N", save = "yes")
#' }
#'
#' @seealso Please also refer to [plotCSCurve()] for regular plots.
#' @export
ggplotCSCurve <- function(data, ..., method = "ml", PEmethod = "fit", colors = NULL,
                          title = "", title_size = 20, title_color = "black", title_face = "bold", title_align = "center",
                          subtitle = "", sub_size = 15, sub_color = "black", sub_face = "italic", sub_align = "center",
                          xlab = "Dose (Gy)", xlab_size = 16, xlab_color = "black", xlab_face = "bold",
                          rem_minor_x = "yes", emph_major_x = "no", major_x_col = "gray", major_x_width = 1,
                          ylab = "Surviving Fraction", ylab_size = 16, ylab_color = "black", ylab_face = "bold",
                          rem_minor_y = "yes", emph_major_y = "no", major_y_col = "gray", major_y_width = 1,
                          legend_title = "", ltitle_size = 20, ltitle_color = "black", ltitle_face = "bold",
                          ltitle_align = "left", legend_pos = "inside", legend_back = "", legend_border = "",
                          legend_border_width = 0,
                          ltext_size = 18, ltext_color = "black", ltext_face = "bold",
                          xtext_size = 14, xtext_color = "black", xtext_face = "bold",
                          ytext_size = 14, ytext_color = "black", ytext_face = "bold",
                          point_shape = 16, point_size = 3.5, segment_type = 1, segment_width = 1,
                          curve_width = 1.5, curve_type = 1, theme = ggplot2::theme_bw(), ylim = NULL, ybreaks = ggplot2::waiver(),
                          save = "no", plot_width = 5, plot_height = 4, units = "in", save_path = getwd(), save_filename = "myplot.pdf") {




  if (!is.null(colors)) {
    if (length(colors) < length(c(...))) {
      warning("Error! The number of colors chosen is less than the number of cell types chosen. It should be at least equal to that!")
    }
  }

  cell_types <- factor(c(...), levels = c(...))
  l <- length(cell_types)
  merged_df <- .dfforggPlot(data, cell_types[1], method, PEmethod)


  if (l > 1) {
    for (i in 2:l) {
      df <- .dfforggPlot(data, cell_types[i], method, PEmethod)
      merged_df <- rbind(merged_df, df)
    }
  }
  dose_0_df <- merged_df %>% dplyr::filter(doses == doses[1])
  alpha_vec <- c(dose_0_df$alpha)
  beta_vec <- c(dose_0_df$beta)
  doses <- unique(data$dose)
  n_seq <- 100
  df_sim <- data.frame(d = rep(with(merged_df, seq(0, max(doses), length.out = n_seq)), l),
                       ctype = rep(cell_types, each = n_seq),
                       alpha = rep(alpha_vec, each = n_seq),
                       beta = rep(beta_vec, each = n_seq)) %>%
    dplyr::mutate(y_hat = exp(alpha*d + beta*(d^2)))
  if (is.null(colors)) {
    colors <- randomcoloR::distinctColorPalette(l)
  }


  p <- ggplot2::ggplot(data = merged_df, ggplot2::aes(x = doses, y = pts, color = ctype)) +
    ggplot2::geom_point(ggplot2::aes(x = doses, y = pts), shape = point_shape, size = point_size) +
    ggplot2::geom_segment(ggplot2::aes(x = doses, xend = doses, y = pts_minus_sems, yend = pts_plus_sems), linetype = segment_type, size = segment_width) +
    ggplot2::geom_line(data = df_sim, ggplot2::aes(y = y_hat, x = d, color = ctype), linetype = curve_type, size = curve_width) +
    ggplot2::scale_x_continuous(breaks = doses) +
    ggplot2::scale_y_log10(limits = ylim, breaks = ybreaks) +
    ggplot2::scale_color_manual(values = colors) +
    ggplot2::labs(title = title, subtitle = subtitle, x = xlab, y = ylab, color = legend_title) +
    theme +
    ggplot2::theme(
      plot.title = ggplot2::element_text(size = title_size, color = title_color, face = title_face),
      plot.subtitle = ggplot2::element_text(size = sub_size, color = sub_color, face = sub_face),
      legend.title = ggplot2::element_text(size = ltitle_size, color = ltitle_color, face = ltitle_face),
      legend.text = ggplot2::element_text(size = ltext_size, color = ltext_color, face = ltext_face),
      axis.title.x = ggplot2::element_text(size = xlab_size, color = xlab_color, face = xlab_face),
      axis.title.y = ggplot2::element_text(size = ylab_size, color = ylab_color, face = ylab_face),
      axis.text.x = ggplot2::element_text(size = xtext_size, color = xtext_color, face = xtext_face),
      axis.text.y = ggplot2::element_text(size = ytext_size, color = ytext_color, face = ytext_face),
      legend.background = ggplot2::element_rect(fill = legend_back, color = legend_border, size = legend_border_width))

  if (title_align == "center") {
    p <- p + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))
  } else if (title_align == "left") {
    p <- p + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0))
  } else {
    p <- p + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 1))
  }

  if (sub_align == "center") {
    p <- p + ggplot2::theme(plot.subtitle = ggplot2::element_text(hjust = 0.5))
  } else if (sub_align == "left") {
    p <- p + ggplot2::theme(plot.subtitle = ggplot2::element_text(hjust = 0))
  } else {
    p <- p + ggplot2::theme(plot.subtitle = ggplot2::element_text(hjust = 1))
  }

  if (ltitle_align == "center") {
    p <- p + ggplot2::theme(legend.title.align = 0.5)
  } else if (ltitle_align == "left") {
    p <- p + ggplot2::theme(legend.title.align = 0)
  } else {
    p <- p + ggplot2::theme(legend.title.align = 1)
  }

  if (title == "") {
    p <- p + ggplot2::theme(plot.title = ggplot2::element_blank())
  }

  if (subtitle == "") {
    p <- p + ggplot2::theme(plot.subtitle = ggplot2::element_blank())
  }

  if (legend_title == "") {
    p <- p + ggplot2::theme(legend.title = ggplot2::element_blank())
  }

  if (xlab == "") {
    p <- p + ggplot2::theme(axis.title.x = ggplot2::element_blank())
  }

  if (ylab == "") {
    p <- p + ggplot2::theme(axis.title.y = ggplot2::element_blank())
  }

  if (legend_pos == "inside") {
    p <- p + ggplot2::theme(legend.justification = c(-0.2, -0.2), legend.position = c(0, 0),
                            legend.background = ggplot2::element_rect(fill = legend_back, color = legend_border,
                                                                      size = legend_border_width),
                            legend.margin = ggplot2::margin(0.2, 0.2, 0.2, 0.2, "cm"),
                            legend.spacing = ggplot2::unit(0.4, "cm"),
                            legend.key = ggplot2::element_blank(),
                            legend.key.size = ggplot2::unit(1.5, "line"))
  }

  if (legend_pos == "none") {
    p <- p + ggplot2::theme(legend.position = "none")
  }

  if (legend_back == "") {
    p <- p + ggplot2::theme(legend.background = ggplot2::element_blank())
  }

  if (rem_minor_x == "yes") {
    p <- p + ggplot2::theme(panel.grid.minor.x = ggplot2::element_blank())
  }

  if (rem_minor_y == "yes") {
    p <- p + ggplot2::theme(panel.grid.minor.y = ggplot2::element_blank())
  }

  if (emph_major_x == "yes") {
    p <- p + ggplot2::theme(panel.grid.major.x = ggplot2::element_line(color = major_x_col, size = major_x_width))
  }

  if (emph_major_y == "yes") {
    p <- p + ggplot2::theme(panel.grid.major.y = ggplot2::element_line(color = major_y_col, size = major_y_width))
  }

  if (save == "yes") {
    ggplot2::ggsave(plot = p, filename = save_filename, path = save_path, width = plot_width, height = plot_height, units = units)
  }

  return(p)
}
